models/auth/token_scope.go

```diff
@@ -0,0 +1,234 @@
+// Copyright 2022 The Gitea Authors. All rights reserved.
+// Use of this source code is governed by a MIT-style
+// license that can be found in the LICENSE file.
+
+package auth
+
+import (
+	"fmt"
+	"strings"
+
+	"code.gitea.io/gitea/modules/util"
+)
+
+// AccessTokenScope represents the scope for an access token.
+type AccessTokenScope string
+
+const (
+	AccessTokenScopeAll = "all"
+
+	AccessTokenScopeRepo       = "repo"
+	AccessTokenScopeRepoStatus = "repo:status"
+	AccessTokenScopePublicRepo = "public_repo"
+
+	AccessTokenScopeAdminOrg = "admin:org"
+	AccessTokenScopeWriteOrg = "write:org"
+	AccessTokenScopeReadOrg  = "read:org"
+
+	AccessTokenScopeAdminPublicKey = "admin:public_key"
+	AccessTokenScopeWritePublicKey = "write:public_key"
+	AccessTokenScopeReadPublicKey  = "read:public_key"
+
+	AccessTokenScopeAdminRepoHook = "admin:repo_hook"
+	AccessTokenScopeWriteRepoHook = "write:repo_hook"
+	AccessTokenScopeReadRepoHook  = "read:repo_hook"
+
+	AccessTokenScopeAdminOrgHook = "admin:org_hook"
+
+	AccessTokenScopeNotification = "notification"
+
+	AccessTokenScopeUser       = "user"
+	AccessTokenScopeReadUser   = "read:user"
+	AccessTokenScopeUserEmail  = "user:email"
+	AccessTokenScopeUserFollow = "user:follow"
+
+	AccessTokenScopeDeleteRepo = "delete_repo"
+
+	AccessTokenScopePackage       = "package"
+	AccessTokenScopeWritePackage  = "write:package"
+	AccessTokenScopeReadPackage   = "read:package"
+	AccessTokenScopeDeletePackage = "delete:package"
+
+	AccessTokenScopeAdminGPGKey = "admin:gpg_key"
+	AccessTokenScopeWriteGPGKey = "write:gpg_key"
+	AccessTokenScopeReadGPGKey  = "read:gpg_key"
+
+	AccessTokenScopeAdminApplication = "admin:application"
+	AccessTokenScopeWriteApplication = "write:application"
+	AccessTokenScopeReadApplication  = "read:application"
+
+	AccessTokenScopeSudo = "sudo"
+)
+
+// AccessTokenScopeBitmap represents a bitmap of access token scopes.
+type AccessTokenScopeBitmap uint64
```

Comment:
Are we sure we only have <64 scopes for all time?

Reply:
May not be indefinitely, but it should be for quite some time
But since bitmap is created only during runtime and is not stored in database, I think bitmap is a good choice here that offers simplicity. When we have >64 scopes, rewrite is needed in `models/auth/token_scope.go` file only and won't impact any other files.
I don't have a good alternative implementation in mind. If there's one, I can change though.

Reply:
So let's leave a comment after `AccessTokenScopeSudo` that when the length > 64, we need to refactor the whole implementation.

Reply:
Another question about this bitmap. Is it safe to add a new entry inbetween existing entries? This will shift the following entries but if the bitmap value is not stored somewhere, it is a safe operation.

Reply:
I also do not think it's a good design by using bitmap. It's not like what the comment said: `refactoring the whole implementation in this file (and only this file) is needed.`, because altering database structure is more difficult in the future. The hacky method would cause problem one day in the future.

If let me decide, the scopes should be refactored before 1.19, and it should make the scopes to be stored by a more flexible and stable approach in database.

Reply:
If I understand it correctly this bitmap is only stored in memory and not persisted in the database. In the database the string version is stored.

Reply:
ps: as KN4CK3R said, if the bitmap would never be used anywhere else, nor be stored in the database , then it's fine.